examples/demo-unused-variable-warner.rkt

```diff
@@ -0,0 +1,106 @@
+                 (collect-ids! #'b phase)]
+        ))
+    
+    (define-pass mark-unused-variable : (FullyExpandedProgram ModuleBeginForm) (prog) -> (FullyExpandedProgram ModuleBeginForm) ()
+      (ModuleBeginForm : ModuleBeginForm (prog phase) -> ModuleBeginForm ()
+                       [(#%plain-module-begin ,s ,[ml* 0 -> ml*] ... ,[ml 0 -> ml])
+                        prog])
+      (Expr : Expr (prog phase) -> Expr ()
+            [,x 
+             (mark-used! x phase)
+             prog]
+            [(#%top ,s ,x)
+             (mark-used! x phase)
+             prog]
+            [(#%variable-reference ,s ,x)
+             (mark-used! x phase)
+             prog]
+            [(#%variable-reference-top ,s ,x)
+             (mark-used! x phase)
+             prog]
+            [(#%plain-lambda ,s ,fml ,[body*] ... ,[body])
+             (collect-ids! fml phase)
+             prog]
+            [(case-lambda ,s [,fml ,[body*] ... ,[body]] ...)
+             (for ([f fml])
+               (collect-ids! f phase))
+             prog]
+            [(let-values ,s ([(,x* ...) ,[e]] ...) ,[body*] ... ,[body])
+             (for ([x x*])
+               (collect-ids! x phase))
+             prog]
+            [(letrec-values ,s ([(,x* ...) ,[e]] ...) ,[body*] ... ,[body])
+             (for ([x x*])
+               (collect-ids! x phase))
+             prog]
+            )
+      (GeneralTopLevelForm : GeneralTopLevelForm (prog phase) -> GeneralTopLevelForm ()
+                           #;[(define-values ,s (,x ...) ,[e])
+                              (collect-ids! x phase)
+                              prog
+                              ] ;;ignore module level definitions
+                           [(define-syntaxes ,s (,x ...) ,[e0 (+ phase 1) -> e1])
+                            prog]
+                           )
+      (ModuleLevelForm : ModuleLevelForm (prog phase) -> ModuleLevelForm ()
+                       [(begin-for-syntax ,s ,[ml* (+ phase 1) -> ml^] ...)
+                        prog])
+      (SubModuleForm : SubModuleForm (prog phase) -> SubModuleForm ()
+                     )
+      (TopLevelForm : TopLevelForm (prog phase) -> SubModuleForm ()
+                    )
+                        
+      (ModuleBeginForm prog 0)
+      )
+    (mark-unused-variable prog)
+    (define unused-set (apply append (for/list ([(ph vars) defined-variables])
+                                       (define same-phase-used (hash-ref used-variables ph #f))
+                                       (when same-phase-used (free-id-set-subtract! vars same-phase-used))
+                                       (free-id-set->list vars)
+                                       )))
+
+    (unless (null? unused-set)
+      (raise-syntax-error 'unused-variable-warner
+                          "warning : unused variables"
+                          #f
+                          #f
+                          unused-set))
+    prog
+    
+    )
+  )
+
```

Comment:
Please adjust the closing parens according to https://docs.racket-lang.org/style/Textual_Matters.html